deepfence_server/pkg/registry/acr/client.go

```diff
@@ -11,39 +12,103 @@ import (
 	"github.com/deepfence/ThreatMapper/deepfence_utils/log"
 )
 
-var client = &http.Client{Timeout: 10 * time.Second}
+const (
+	PerPageCount       = 100
+	ParallelImageFetch = 10
+)
+
+var (
+	client                = &http.Client{Timeout: 10 * time.Second}
+	parllelImageProcessor *tunny.Pool
+	queue                 chan []model.IngestedContainerImage
+)
+
+func init() {
+	parllelImageProcessor = tunny.NewFunc(ParallelImageFetch, fetchImageWithTags)
+	queue = make(chan []model.IngestedContainerImage)
+}
+
+type RepoDetails struct {
+	Url        string
+	UserName   string
+	Password   string
+	NameSpace  string
+	Repository string
+}
 
 func listImagesRegistryV2(url, namespace, userName,
 	password string) ([]model.IngestedContainerImage, error) {
 	var images []model.IngestedContainerImage
 
-	repos, err := listCatalogRegistryV2(url, namespace, userName, password)
+	repos, err := getRepos(url, userName, password)
 	if err != nil {
 		log.Error().Msg(err.Error())
 		return nil, err
 	}
 	for _, repo := range repos {
-		repoTags, err := listRepoTagsV2(url, namespace, userName, password, repo)
-		if err != nil {
-			log.Error().Msg(err.Error())
-			continue
+		r := RepoDetails{
+			Url:        url,
+			UserName:   userName,
+			Password:   password,
+			NameSpace:  namespace,
+			Repository: repo,
+		}
+		go parllelImageProcessor.Process(r)
+	}
+	for _, _ = range repos {
+		select {
+		case t := <-queue:
+			images = append(images, t...)
 		}
-		log.Debug().Msgf("tags for image %s/%s are %s", repo, repoTags.Name, repoTags.Tags)
-
-		images = append(images, getImageWithTags(url, namespace, userName, password, repo, repoTags)...)
 	}
 
 	return images, nil
 }
 
-func listCatalogRegistryV2(url, namespace, userName,
-	password string) ([]string, error) {
+func fetchImageWithTags(rInterface interface{}) interface{} {
+	r, ok := rInterface.(*RepoDetails)
+	if !ok {
+		log.Error().Msg("Error processing repo details")
+		queue <- []model.IngestedContainerImage{}
```

Comment:
We should create a variable `images` at top 
```
images := model.IngestedContainerImage{}
```
And use the `defer queue <- images` 
Reason being, if later we forget to add the image slice to the channel, the consumer side will be blocked forever.